lib/ansible/modules/cloud/opennebula/one_service.py

```diff
@@ -0,0 +1,762 @@
+    except Exception as e:
+        module.fail_json(msg=str(e))
+
+    return json.load(response)
+
+
+def get_service(module, auth, pred):
+    all_services_dict = get_all_services(module, auth)
+
+    found = 0
+    found_service = None
+    service_name = ''
+
+    if "DOCUMENT_POOL" in all_services_dict and "DOCUMENT" in all_services_dict["DOCUMENT_POOL"]:
+        for service in all_services_dict["DOCUMENT_POOL"]["DOCUMENT"]:
+            if pred(service):
+                found = found + 1
+                found_service = service
+                service_name = service["NAME"]
+
+    # fail if there are more services with same name
+    if found > 1:
+        module.fail_json(msg="There are multiple services with a name: '" +
+                         service_name + "'. You have to use a unique service name or use 'service_id' instead.")
+    elif found <= 0:
+        return None
+    else:
+        return found_service
+
+
+def get_service_by_id(module, auth, service_id):
+    return get_service(module, auth, lambda service: (int(service["ID"]) == int(service_id))) if service_id else None
+
+
+def get_service_by_name(module, auth, service_name):
+    return get_service(module, auth, lambda service: (service["NAME"] == service_name))
+
+
+def get_service_info(module, auth, service):
+
+    result = {
+        "service_id": int(service["ID"]),
+        "service_name": service["NAME"],
+        "group_id": int(service["GID"]),
+        "group_name": service["GNAME"],
+        "owner_id": int(service["UID"]),
+        "owner_name": service["UNAME"],
+        "state": STATES[service["TEMPLATE"]["BODY"]["state"]]
+    }
+
+    roles_status = service["TEMPLATE"]["BODY"]["roles"]
+    roles = []
+    for role in roles_status:
+        nodes_ids = []
+        if "nodes" in role:
+            for node in role["nodes"]:
+                nodes_ids.append(node["deploy_id"])
+        roles.append({"name": role["name"], "cardinality": role["cardinality"], "state": STATES[int(role["state"])], "ids": nodes_ids})
+
+    result["roles"] = roles
+    result["mode"] = int(parse_service_permissions(service))
+
+    return result
+
+
+def create_service(module, auth, template_id, service_name, custom_attrs, unique, wait, wait_timeout):
+    # make sure that the values in custom_attrs dict are strings
+    custom_attrs_with_str = dict((k, str(v)) for k, v in custom_attrs.items())
+
+    data = {
+        "action": {
+            "perform": "instantiate",
+            "params": {
+                "merge_template": {
+                    "custom_attrs_values": custom_attrs_with_str,
+                    "name": service_name
+                }
+            }
+        }
+    }
+
+    try:
+        response = open_url(auth.url + "/service_template/" + str(template_id) + "/action", method="POST",
+                            data=json.dumps(data), force_basic_auth=True, url_username=auth.user, url_password=auth.password)
```

Comment:
not a blocker but there is also a helper for json.dumps https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/basic.py#L792